File: src/main/java/jb/integrator/MedianIntegrator.java
package jb.integrator;

import de.bwaldvogel.liblinear.Model;
import jb.config.Opts;
import jb.data.Dataset;
import jb.data.FunctionValue;
import jb.data.IntegratedModel;
import jb.data.SelectedTuple;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import static jb.util.MathUtils.arePermutations;
import static jb.util.ModelUtils.*;

public class MedianIntegrator implements Integrator {
    @Override
    public IntegratedModel integrate(SelectedTuple selectedTuple, List<Model> clfs, Dataset dataset, Opts opts) {

        IntegratedModel duplicatedModel = getDuplicatedModel(selectedTuple, clfs, dataset, opts);
        List<Double> resultAs = new ArrayList<>();
        List<Double> resultBs = new ArrayList<>();
        List<Double> resultX = new ArrayList<>();
        resultAs.add(duplicatedModel.getA()[0]);
        resultBs.add(duplicatedModel.getB()[0]);
        resultX.add(duplicatedModel.getX()[0]);
        for (int i = 0; i < duplicatedModel.getX().length - 2; i++) {
            if (duplicatedModel.getA()[i] == duplicatedModel.getA()[i + 1] && duplicatedModel.getB()[i] == duplicatedModel.getB()[i + 1]) continue;
            resultAs.add(duplicatedModel.getA()[i + 1]);
            resultBs.add(duplicatedModel.getB()[i + 1]);
            resultX.add(duplicatedModel.getX()[i + 1]);
        }
        resultX.add(duplicatedModel.getX()[duplicatedModel.getX().length - 1]);
        return IntegratedModel.builder().
                a(resultAs.stream().mapToDouble(Double::doubleValue).toArray()).
                b(resultBs.stream().mapToDouble(Double::doubleValue).toArray()).
                x(resultX.stream().mapToDouble(Double::doubleValue).toArray()).build();
    }

    private IntegratedModel getDuplicatedModel(SelectedTuple selectedTuple, List<Model> clfs, Dataset dataset, Opts opts) {
        List<Double> duplicatingAs = new ArrayList<>();
        List<Double> duplicatingBs = new ArrayList<>();
        List<Double> duplicatingX = new ArrayList<>();
        for (int i = 0; i < opts.getNumberOfSpaceParts(); i++) {
            List<Model> selectedClfs = pickModels(clfs, selectedTuple.getIndices()[i]);
            double[] as = getAs(selectedClfs);
            double[] bs = getBs(selectedClfs);
            List<Double> crosspoints = getAllCrosspoints(as, bs);
            List<Double> filteredCrosspoints = filterCrosspoints(crosspoints, dataset, i, opts);
            filteredCrosspoints.sort(Double::compare);
            int[] previousIndices = null;
            for (int j = 0; j < filteredCrosspoints.size() - 1; j++) {
                int[] indices = getMiddleIndices(as, bs, filteredCrosspoints, j);
                if (sameIndices(previousIndices, indices)) continue;
                duplicatingAs.add(indices.length == 1 ? as[indices[0]] : (as[indices[0]] + as[indices[1]]) / 2);
                duplicatingBs.add(indices.length == 1 ? bs[indices[0]] : (bs[indices[0]] + bs[indices[1]]) / 2);
                duplicatingX.add(filteredCrosspoints.get(j));
                previousIndices = indices;
            }
        }
        duplicatingX.add(dataset.getMaxX());
        return IntegratedModel.builder().
                a(duplicatingAs.stream().mapToDouble(Double::doubleValue).toArray()).
                b(duplicatingBs.stream().mapToDouble(Double::doubleValue).toArray()).
                x(duplicatingX.stream().mapToDouble(Double::doubleValue).toArray()).build();
    }

    private int[] getMiddleIndices(double[] as, double[] bs, List<Double> filteredCrosspoints, int j) {
        List<FunctionValue> functionValues = new ArrayList<>();
        for (int k = 0; k < as.length; k++) {
            functionValues.add(new FunctionValue(k, as[k] * (filteredCrosspoints.get(j) + filteredCrosspoints.get(j + 1)) / 2 + bs[k]));
        }
        functionValues.sort(Comparator.comparingDouble(FunctionValue::getValue));
        return getMiddleIndices(functionValues);
    }

    private boolean sameIndices(int[] previousIndices, int[] indices) {
        return previousIndices != null && previousIndices.length == indices.length && arePermutations(previousIndices, indices);
    }

    private int[] getMiddleIndices(List<FunctionValue> functionValues) {
        return functionValues.size() % 2 == 0 ?
                new int[]{functionValues.get(functionValues.size() / 2 - 1).getIndex(), functionValues.get(functionValues.size() / 2).getIndex()} :
                new int[]{functionValues.get((functionValues.size() - 1) / 2).getIndex()};
    }

    private List<Double> getAllCrosspoints(double[] as, double[] bs) {
        List<Double> x = new ArrayList<>();
        for (int i = 0; i < as.length; i++) {
            for (int j = 0; j < bs.length; j++) {
                if (i == j || as[i] == as[j]) continue;
                x.add((bs[i] - bs[j]) / (as[i] - as[j]));
            }
        }
        return x;
    }

    private List<Double> filterCrosspoints(List<Double> crosspoints, Dataset dataset, int index, Opts opts) {
        List<Double> filteredCrosspoints = new ArrayList<>();
        filteredCrosspoints.add(getLowerLimit(dataset, index, opts));
        filteredCrosspoints.add(getHigherLimit(dataset, index, opts));
        for (Double crosspoint : crosspoints) {
            if (crosspoint < getLowerLimit(dataset, index, opts) || crosspoint > getHigherLimit(dataset, index, opts) || filteredCrosspoints.contains(crosspoint))
                continue;
            filteredCrosspoints.add(crosspoint);
        }
        return filteredCrosspoints;
    }

    private double getLowerLimit(Dataset dataset, int index, Opts opts) {
        return dataset.getMinX() + index * (dataset.getMaxX() - dataset.getMinX()) / opts.getNumberOfSpaceParts();
    }

    private double getHigherLimit(Dataset dataset, int index, Opts opts) {
        return dataset.getMinX() + (index + 1) * (dataset.getMaxX() - dataset.getMinX()) / opts.getNumberOfSpaceParts();
    }

}

File: src/main/java/jb/integrator/MedianIntegrator.kt
package jb.integrator

import de.bwaldvogel.liblinear.Model
import jb.config.Opts
import jb.data.Dataset
import jb.data.FunctionValue
import jb.data.IntegratedModel
import jb.data.SelectedTuple


import jb.util.MathUtils.arePermutations
import jb.util.ModelUtils.*

class MedianIntegrator : Integrator {
    override fun integrate(selectedTuple: SelectedTuple, clfs: List<Model>, dataset: Dataset, opts: Opts): IntegratedModel {

        val (a, b, x) = getDuplicatedModel(selectedTuple, clfs, dataset, opts)
        val resultAs = ArrayList<Double>()
        val resultBs = ArrayList<Double>()
        val resultX = ArrayList<Double>()
        resultAs.add(a[0])
        resultBs.add(b[0])
        resultX.add(x[0])
        for (i in 0 until x.size - 2) {
            if (a[i] == a[i + 1] && b[i] == b[i + 1]) continue
            resultAs.add(a[i + 1])
            resultBs.add(b[i + 1])
            resultX.add(x[i + 1])
        }
        resultX.add(x[x.size - 1])
        /*return IntegratedModel(
                resultAs.stream().mapToDouble(ToDoubleFunction<Double> { it.toDouble() }).toArray(),
                resultBs.stream().mapToDouble(ToDoubleFunction<Double> { it.toDouble() }).toArray(),
                resultX.stream().mapToDouble(ToDoubleFunction<Double> { it.toDouble() }).toArray())*/
        return IntegratedModel(resultAs.toDoubleArray(), resultBs.toDoubleArray(), resultX.toDoubleArray())
    }

    private fun getDuplicatedModel(selectedTuple: SelectedTuple, clfs: List<Model>, dataset: Dataset, opts: Opts): IntegratedModel {
        val duplicatingAs = ArrayList<Double>()
        val duplicatingBs = ArrayList<Double>()
        val duplicatingX = ArrayList<Double>()
        for (i in 0 until opts.numberOfSpaceParts) {
            val selectedClfs = pickModels(clfs, selectedTuple.indices[i])
            val `as` = getAs(selectedClfs)
            val bs = getBs(selectedClfs)
            val crosspoints = getAllCrosspoints(`as`, bs)
            val filteredCrosspoints = filterCrosspoints(crosspoints, dataset, i, opts)
            filteredCrosspoints.sorted()
            var previousIndices: IntArray? = null
            for (j in 0 until filteredCrosspoints.size - 1) {
                val indices = getMiddleIndices(`as`, bs, filteredCrosspoints, j)
                if (sameIndices(previousIndices, indices)) continue
                duplicatingAs.add(if (indices.size == 1) `as`[indices[0]] else (`as`[indices[0]] + `as`[indices[1]]) / 2)
                duplicatingBs.add(if (indices.size == 1) bs[indices[0]] else (bs[indices[0]] + bs[indices[1]]) / 2)
                duplicatingX.add(filteredCrosspoints[j])
                previousIndices = indices
            }
        }
        duplicatingX.add(dataset.maxX)
        return IntegratedModel(duplicatingAs.toDoubleArray(), duplicatingBs.toDoubleArray(), duplicatingX.toDoubleArray())
    }

    private fun getMiddleIndices(`as`: DoubleArray, bs: DoubleArray, filteredCrosspoints: List<Double>, j: Int): IntArray {
        val functionValues = ArrayList<FunctionValue>()
        for (k in `as`.indices) {
            functionValues.add(FunctionValue(k, `as`[k] * (filteredCrosspoints[j] + filteredCrosspoints[j + 1]) / 2 + bs[k]))
        }
        functionValues.sortedBy { it.value }
        return getMiddleIndices(functionValues)
    }

    private fun sameIndices(previousIndices: IntArray?, indices: IntArray): Boolean {
        return previousIndices != null && previousIndices.size == indices.size && arePermutations(previousIndices, indices)
    }

    private fun getMiddleIndices(functionValues: List<FunctionValue>): IntArray {
        return if (functionValues.size % 2 == 0)
            intArrayOf(functionValues[functionValues.size / 2 - 1].index, functionValues[functionValues.size / 2].index)
        else
            intArrayOf(functionValues[(functionValues.size - 1) / 2].index)
    }

    private fun getAllCrosspoints(`as`: DoubleArray, bs: DoubleArray): List<Double> {
        val x = ArrayList<Double>()
        for (i in `as`.indices) {
            for (j in bs.indices) {
                if (i == j || `as`[i] == `as`[j]) continue
                x.add((bs[i] - bs[j]) / (`as`[i] - `as`[j]))
            }
        }
        return x
    }

    private fun filterCrosspoints(crosspoints: List<Double>, dataset: Dataset, index: Int, opts: Opts): List<Double> {
        val filteredCrosspoints = ArrayList<Double>()
        filteredCrosspoints.add(getLowerLimit(dataset, index, opts))
        filteredCrosspoints.add(getHigherLimit(dataset, index, opts))
        for (crosspoint in crosspoints) {
            if (crosspoint < getLowerLimit(dataset, index, opts) || crosspoint > getHigherLimit(dataset, index, opts) || filteredCrosspoints.contains(crosspoint))
                continue
            filteredCrosspoints.add(crosspoint)
        }
        return filteredCrosspoints
    }

    private fun getLowerLimit(dataset: Dataset, index: Int, opts: Opts): Double {
        return dataset.minX + index * (dataset.maxX - dataset.minX) / opts.numberOfSpaceParts
    }

    private fun getHigherLimit(dataset: Dataset, index: Int, opts: Opts): Double {
        return dataset.minX + (index + 1) * (dataset.maxX - dataset.minX) / opts.numberOfSpaceParts
    }

}
